designs/0004-sparse-matrices.md

```diff
@@ -0,0 +1,387 @@
+
+```cpp
+template <typename Derived1, typename Derived2>
+eigen_return_t<Derived1, Derived2> add(const Eigen::EigenBase<Derived1>& A_,
+                                       const Eigen::EigenBase<Derived2>& B_) {
+    // Pull out the Derived type                              
+    Derived1 const& A = A_.derived();
+    Derived2 const& B = B_.derived();
+    return ((A+B).eval()).transpose();
+}
+```
+
+Where `return_derived_t` would deduce the correct derived return type based on the `Scalar` value of the `Derived*` types. This is nice because for places where Eigen supports both dense and sparse operations we do not have to duplicate code. For places where the sparse and dense operations differ we can have sparse matrix template specializations. There has been a lot of discussion on this refactor in the past (see [this](https://groups.google.com/forum/#!topic/stan-dev/ZKYCQ3Y7eY0) Google groups post and [this](https://github.com/stan-dev/math/issues/62) issue). Though looking at the two forms it seems like using `A.coeff()` for access instead of `operator()` would be sufficient to handle the coefficient access error Dan saw.
+
+### The Simple Way
+
+If we would rather not refactor the math library we can keep our current templates and have specializations for Sparse matrices.
+```cpp
+template <typename T1, typename T2>
+inline Eigen::SparseMatrixBase<return_type_t<T1, T2>> add(
+    const Eigen::SparseMatrixBase<T1>& m1, const Eigen::SparseMatrixBase<T2>& m2) {
+  check_matching_dims("add", "m1", m1, "m2", m2);
+  return m1 + m2;
+}
+```
+
+### Keeping Permutation Matrix from Cholesky
+
+`SimplicalCholeskybase` keeps the permutation matrix, when the user uses a `sparse_cholesky_matrix` we can pull out this permutation matrix and keep it to use in the next iteration. We do this through `EIGEN_SPARSEMATRIX_BASE_PLUGIN`, adding the permutation matrix to the input matrix. This adds a bito of state, but assuming the sparse matrices are fixed in size and sparsity this should be fine.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Doing this improperly could lead to serious amounts of "code smell" aka duplicated code, confusing templates, etc.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Why is this design the best in the space of possible designs?
+
+This is the most full-feature design I can think of that makes sparse matrices simple for stan users while giving the stan-math library a lot of flexibility.
+
+- What other designs have been considered and what is the rationale for not choosing them?
+
+Instead of the full feature spec like the above we could have sparse_matrix only be a result type
+
+```stan
+parameters {
+  real theta_raw[K];
+...
+model {
+  sparse_matrix[M, N] theta = to_sparse_matrix(M, N, mm, nn, theta_raw);
+  ...
+```
+
+Or alternativly we can extend the `csr_*` functions, though this requires a lot of recomputation.
```

Comment:
The `csr_*` functions are not good and should go away.  Also it's inconsistent to have row oriented sparse-matrices and column oriented dense matrices.